### PR TITLE
Refactor to use linear device to represent backstore

### PR DESCRIPTION
**Pull request by mulkieran**
See #722 for prior discussion.

### CODE REVIEW

**Comment by mulkieran** [comment]
@agrover review, please?

**Comment by agrover** [comment]
@trgill @tasleson if you have any review comments, please get them in today.

File: src/engine/strat_engine/backstore/store.rs [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+// Code to handle the backing store of a pool.
+
+use std::collections::HashMap;
+use std::path::{Path, PathBuf};
+
+use chrono::{DateTime, Utc};
+
+use devicemapper::{Device, DM, DmDevice, LinearDev, Sectors};
```
Comment (tasleson) [inline src/engine/strat_engine/backstore/store.rs]: Please sort [`DM, Device, DmDevice, LinearDev, Sectors`]

File: src/engine/strat_engine/backstore/store.rs [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+
+use super::super::super::engine::BlockDev;
+use super::super::super::errors::{EngineError, EngineResult, ErrorEnum};
+use super::super::super::types::{DevUuid, PoolUuid};
+
+use super::super::dmnames::{CacheRole, format_backstore_name};
+use super::super::serde_structs::{Recordable, BackstoreSave};
```
Comment (tasleson) [inline src/engine/strat_engine/backstore/store.rs]: Please sort `BackstoreSave, Recordable`

File: src/engine/strat_engine/thinpool/thinpool.rs [hunk 3]
```diff
@@ -11,23 +11,25 @@ use std::process::Command;
...
 use super::super::super::structures::Table;
-use super::super::super::types::{DevUuid, FilesystemUuid, Name, PoolUuid, RenameAction};
+use super::super::super::types::{FilesystemUuid, Name, PoolUuid, RenameAction};
 
-use super::super::blockdevmgr::{BlkDevSegment, BlockDevMgr, Segment, map_to_dm};
+use super::super::backstore::Backstore;
 use super::super::device::wipe_sectors;
 use super::super::devlinks;
+use super::super::dmnames::{FlexRole, ThinPoolRole, ThinRole, format_flex_name,
+                            format_thinpool_name, format_thin_name};
```
Comment (tasleson) [inline src/engine/strat_engine/thinpool/thinpool.rs]: Please sort `FlexRole, ThinPoolRole, ThinRole, format_flex_name, format_thin_name, format_thinpool_name`

File: src/engine/strat_engine/device.rs [hunk 4]
```diff
@@ -4,26 +4,14 @@
 
 // Functions for dealing with devices.
 
-use std::collections::HashMap;
-use std::fs::{File, OpenOptions};
 use std::io::{BufWriter, Seek, SeekFrom, Write};
-use std::os::unix::prelude::AsRawFd;
+use std::fs::OpenOptions;
```
Comment (tasleson) [inline src/engine/strat_engine/device.rs]: `use std::fs::OpenOptions` should come before `use std::io`

File: src/engine/strat_engine/backstore/cleanup.rs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+/// If an error occurs while wiping a blockdev, attempt to wipe all remaining.
+pub fn wipe_blockdevs(blockdevs: &[StratBlockDev]) -> EngineResult<()> {
+    let mut unerased_devnodes = Vec::new();
+
+    for bd in blockdevs {
+        let bd_devnode = bd.devnode.to_owned();
+        bd.wipe_metadata()
+            .unwrap_or_else(|_| unerased_devnodes.push(bd_devnode));
+    }
+
```
Comment (tasleson) [inline src/engine/strat_engine/backstore/cleanup.rs]: Would this be more rustic?

```rust
blockdevs
        .iter()
        .for_each(|bd| {
                      let bd_devnode = bd.devnode.to_owned();
                      bd.wipe_metadata()
                          .unwrap_or_else(|_| unerased_devnodes.push(bd_devnode))
                  });
```

**Comment by tasleson** [comment]
@mulkieran Next time please put re-organize code in a separate commit, so that it would be easier to distinguish moved vs. newly written code.

File: src/engine/strat_engine/backstore/store.rs (hunk 1 above)
Comment (mulkieran) [reply src/engine/strat_engine/backstore/store.rs]: Done.

File: src/engine/strat_engine/backstore/store.rs (hunk 2 above)
Comment (mulkieran) [reply src/engine/strat_engine/backstore/store.rs]: Done.

File: src/engine/strat_engine/thinpool/thinpool.rs (hunk 3 above)
Comment (mulkieran) [reply src/engine/strat_engine/thinpool/thinpool.rs]: ASCII code for "_" precedies code for "p". Should stay as is.

File: src/engine/strat_engine/backstore/cleanup.rs (hunk 5 above)
Comment (mulkieran) [reply src/engine/strat_engine/backstore/cleanup.rs]: Sort of, but there is some feeling that a for is better when the whole purpose is not to compute something but to have the side-effect. So we generally stick with the for-loop. Also, I just moved this code, I didn't work on it in this PR, so I'm going to leave it as is.

File: src/engine/strat_engine/thinpool/thinpool.rs (hunk 3 above)
Comment (tasleson) [reply src/engine/strat_engine/thinpool/thinpool.rs]: @mulkieran Yes `_` (95d) comes before `p` (112d), thus `format_thin_name` comes before `format_thinpool_name` in ASCII sorting.

As I stated in a previous PR I'm not doing these checks by hand anymore.  If we don't like ASCII sort order we can change it, but we need to pick one and stick with it as the tool I wrote will keep flagging an error if `format_thinpool_name` comes before `format_thin_name`.

Comment (mulkieran) [reply src/engine/strat_engine/thinpool/thinpool.rs]: @tasleson When will your tool be mature enough to include in our CI?

Comment (tasleson) [reply src/engine/strat_engine/thinpool/thinpool.rs]: @mulkieran After it gets some use from multiple users and we find the results meet our expectations.

Comment (mulkieran) [reply src/engine/strat_engine/thinpool/thinpool.rs]: @tasleson What are the expectations that you think it doesn't meet yet?

**Pull request merged**